File: apps/shared/src/utils/interest.ts
/**
 * Interest calculation utilities for DeFi RWA platform
 * Matches on-chain Rust contract implementation for consistency
 */

/** Precision for fixed-point calculations (18 decimals) */
export const PRECISION = BigInt("1000000000000000000");

/** Seconds per year for APY calculations */
export const SECONDS_PER_YEAR = 31_536_000;

/** Days per year for interest calculations */
export const DAYS_PER_YEAR = 365;

/**
 * Interest rate model parameters
 * Uses kinked model: rate changes slope at optimal utilization
 */
export interface InterestRateModel {
  /** Base rate (in PRECISION units, e.g., 2% = 0.02 * PRECISION) */
  baseRate: bigint;
  /** Slope below optimal utilization */
  slope1: bigint;
  /** Slope above optimal utilization */
  slope2: bigint;
  /** Optimal utilization rate (e.g., 80% = 0.8 * PRECISION) */
  optimalUtilization: bigint;
}

/**
 * Default interest rate model
 * Base: 2%, Slope1: 4%, Slope2: 75%, Optimal: 80%
 */
export const DEFAULT_INTEREST_RATE_MODEL: InterestRateModel = {
  baseRate: BigInt("20000000000000000"), // 2%
  slope1: BigInt("40000000000000000"), // 4%
  slope2: BigInt("750000000000000000"), // 75%
  optimalUtilization: BigInt("800000000000000000"), // 80%
};

/**
 * Calculate simple interest
 * Formula: principal * rate * time
 * @param principal - Principal amount
 * @param rate - Annual interest rate (decimal, e.g., 0.05 for 5%)
 * @param timeInYears - Time period in years
 * @returns Interest amount
 */
export function calculateSimpleInterest(
  principal: number,
  rate: number,
  timeInYears: number,
): number {
  return principal * rate * timeInYears;
}

/**
 * Calculate compound interest
 * Formula: principal * (1 + rate/n)^(n*time) - principal
 * @param principal - Principal amount
 * @param rate - Annual interest rate (decimal, e.g., 0.05 for 5%)
 * @param timeInYears - Time period in years
 * @param compoundingFrequency - Number of times interest compounds per year
 * @returns Total amount (principal + interest)
 */
export function calculateCompoundInterest(
  principal: number,
  rate: number,
  timeInYears: number,
  compoundingFrequency: number = 1,
): number {
  const ratePerPeriod = rate / compoundingFrequency;
  const totalPeriods = compoundingFrequency * timeInYears;
  return principal * Math.pow(1 + ratePerPeriod, totalPeriods);
}

/**
 * Convert APR to APY
 * Formula: (1 + APR/n)^n - 1
 * @param apr - Annual Percentage Rate (decimal, e.g., 0.05 for 5%)
 * @param compoundingFrequency - Number of times interest compounds per year
 * @returns APY (Annual Percentage Yield)
 */
export function aprToApy(
  apr: number,
  compoundingFrequency: number = 1,
): number {
  return Math.pow(1 + apr / compoundingFrequency, compoundingFrequency) - 1;
}

/**
 * Convert APY to APR
 * Formula: n * ((1 + APY)^(1/n) - 1)
 * @param apy - Annual Percentage Yield (decimal, e.g., 0.05 for 5%)
 * @param compoundingFrequency - Number of times interest compounds per year
 * @returns APR (Annual Percentage Rate)
 */
export function apyToApr(
  apy: number,
  compoundingFrequency: number = 1,
): number {
  return (
    compoundingFrequency * (Math.pow(1 + apy, 1 / compoundingFrequency) - 1)
  );
}

/**
 * Calculate utilization rate
 * Formula: totalBorrows / totalSupply
 * @param totalBorrows - Total borrowed amount
 * @param totalSupply - Total supplied amount
 * @returns Utilization rate in PRECISION units (0 to PRECISION)
 */
export function calculateUtilizationRate(
  totalBorrows: bigint,
  totalSupply: bigint,
): bigint {
  if (totalSupply === BigInt(0)) {
    return BigInt(0);
  }
  return (totalBorrows * PRECISION) / totalSupply;
}

/**
 * Calculate borrow rate using kinked interest rate model
 * Matches Rust contract implementation
 * @param utilization - Utilization rate in PRECISION units
 * @param model - Interest rate model parameters
 * @returns Borrow rate in PRECISION units
 */
export function calculateBorrowRate(
  utilization: bigint,
  model: InterestRateModel = DEFAULT_INTEREST_RATE_MODEL,
): bigint {
  if (utilization <= model.optimalUtilization) {
    // Below optimal: base + utilization * slope1 / optimal
    const slopeComponent =
      (utilization * model.slope1) / model.optimalUtilization;
    return model.baseRate + slopeComponent;
  } else {
    // Above optimal: rate_at_optimal + (utilization - optimal) * slope2 / (1 - optimal)
    const rateAtOptimal = model.baseRate + model.slope1;
    const excessUtilization = utilization - model.optimalUtilization;
    const remaining = PRECISION - model.optimalUtilization;
    const excessComponent = (excessUtilization * model.slope2) / remaining;
    return rateAtOptimal + excessComponent;
  }
}

/**
 * Calculate supply rate from borrow rate
 * Formula: borrow_rate * utilization * (1 - reserve_factor)
 * @param borrowRate - Borrow rate in PRECISION units
 * @param utilization - Utilization rate in PRECISION units
 * @param reserveFactor - Reserve factor in PRECISION units (e.g., 10% = 0.1 * PRECISION)
 * @returns Supply rate in PRECISION units
 */
export function calculateSupplyRate(
  borrowRate: bigint,
  utilization: bigint,
  reserveFactor: bigint,
): bigint {
  const effectiveRate = (borrowRate * utilization) / PRECISION;
  const factor = PRECISION - reserveFactor;
  return (effectiveRate * factor) / PRECISION;
}

/**
 * Calculate projected earnings over time
 * @param principal - Principal amount
 * @param annualRate - Annual rate (decimal, e.g., 0.05 for 5%)
 * @param days - Number of days
 * @returns Projected earnings
 */
export function calculateProjectedEarnings(
  principal: number,
  annualRate: number,
  days: number,
): number {
  const timeInYears = days / DAYS_PER_YEAR;
  return calculateSimpleInterest(principal, annualRate, timeInYears);
}

/**
 * Calculate new interest index based on time elapsed
 * Matches Rust contract implementation
 * @param currentIndex - Current interest index in PRECISION units
 * @param borrowRate - Borrow rate in PRECISION units
 * @param timeElapsed - Time elapsed in seconds
 * @returns New interest index in PRECISION units
 */
export function calculateNewIndex(
  currentIndex: bigint,
  borrowRate: bigint,
  timeElapsed: number,
): bigint {
  if (timeElapsed === 0) {
    return currentIndex;
  }

  // Calculate rate per second
  const ratePerSecond = borrowRate / BigInt(SECONDS_PER_YEAR);

  // Calculate accumulated interest: index * (1 + rate * time)
  const rateTimesTime = ratePerSecond * BigInt(timeElapsed);
  const interestFactor = PRECISION + rateTimesTime;

  return (currentIndex * interestFactor) / PRECISION;
}

/**
 * Convert rate from PRECISION units to decimal
 * @param rate - Rate in PRECISION units
 * @returns Rate as decimal (e.g., 0.05 for 5%)
 */
export function rateToDecimal(rate: bigint): number {
  return Number(rate) / Number(PRECISION);
}

/**
 * Convert decimal rate to PRECISION units
 * @param rate - Rate as decimal (e.g., 0.05 for 5%)
 * @returns Rate in PRECISION units
 */
export function decimalToRate(rate: number): bigint {
  return BigInt(Math.floor(rate * Number(PRECISION)));
}

/**
 * Convert rate to percentage string
 * @param rate - Rate in PRECISION units
 * @param decimals - Number of decimal places (default: 2)
 * @returns Percentage string (e.g., "5.00%")
 */
export function rateToPercentage(rate: bigint, decimals: number = 2): string {
  const decimal = rateToDecimal(rate);
  return `${(decimal * 100).toFixed(decimals)}%`;
}

File: apps/contracts/contracts/defi-rwa/src/lending/interest.rs
use soroban_sdk::{contracttype, Env, String};

use super::keys::{lending_bump, LendingKey};

/// Precision for fixed-point calculations (18 decimals)
pub const PRECISION: i128 = 1_000_000_000_000_000_000;

/// Seconds per year for APY calculations
pub const SECONDS_PER_YEAR: u64 = 31_536_000;

/// Interest rate model parameters
/// Uses linear model: rate = base + (utilization * slope)
#[derive(Clone)]
#[contracttype]
pub struct InterestRateModel {
    /// Base rate (in PRECISION units, e.g., 2% = 0.02 * PRECISION)
    pub base_rate: i128,
    /// Slope below optimal utilization
    pub slope1: i128,
    /// Slope above optimal utilization
    pub slope2: i128,
    /// Optimal utilization rate (e.g., 80% = 0.8 * PRECISION)
    pub optimal_utilization: i128,
}

impl Default for InterestRateModel {
    /// Create default interest rate model
    /// Base: 2%, Slope1: 4%, Slope2: 75%, Optimal: 80%
    fn default() -> Self {
        Self {
            base_rate: 20_000_000_000_000_000,            // 2%
            slope1: 40_000_000_000_000_000,               // 4%
            slope2: 750_000_000_000_000_000,              // 75%
            optimal_utilization: 800_000_000_000_000_000, // 80%
        }
    }
}

impl InterestRateModel {
    /// Calculate borrow rate based on utilization
    /// Uses checked arithmetic to prevent overflow
    pub fn calculate_borrow_rate(&self, utilization: i128) -> i128 {
        if utilization <= self.optimal_utilization {
            // Below optimal: base + utilization * slope1 / optimal
            let slope_component = utilization
                .checked_mul(self.slope1)
                .expect("Borrow rate overflow: utilization * slope1")
                / self.optimal_utilization;
            self.base_rate
                .checked_add(slope_component)
                .expect("Borrow rate overflow: base + slope_component")
        } else {
            // Above optimal: rate_at_optimal + (utilization - optimal) * slope2 / (1 - optimal)
            let rate_at_optimal = self
                .base_rate
                .checked_add(self.slope1)
                .expect("Borrow rate overflow: base + slope1");
            let excess_utilization = utilization - self.optimal_utilization;
            let remaining = PRECISION - self.optimal_utilization;
            let excess_component = excess_utilization
                .checked_mul(self.slope2)
                .expect("Borrow rate overflow: excess * slope2")
                / remaining;
            rate_at_optimal
                .checked_add(excess_component)
                .expect("Borrow rate overflow: rate_at_optimal + excess_component")
        }
    }

    /// Calculate supply rate based on borrow rate and utilization
    /// Uses checked arithmetic to prevent overflow
    pub fn calculate_supply_rate(
        &self,
        borrow_rate: i128,
        utilization: i128,
        reserve_factor: i128,
    ) -> i128 {
        // supply_rate = borrow_rate * utilization * (1 - reserve_factor)
        let effective_rate = borrow_rate
            .checked_mul(utilization)
            .expect("Supply rate overflow: borrow_rate * utilization")
            / PRECISION;
        let factor = PRECISION - reserve_factor;
        effective_rate
            .checked_mul(factor)
            .expect("Supply rate overflow: effective_rate * factor")
            / PRECISION
    }
}

/// Interest rate storage helpers
pub struct InterestStorage;

impl InterestStorage {
    /// Store interest rate model for pool
    pub fn set_model(env: &Env, pool_id: &String, model: &InterestRateModel) {
        let key = LendingKey::InterestRateModel(pool_id.clone());
        env.storage().persistent().set(&key, model);
        env.storage().persistent().extend_ttl(
            &key,
            lending_bump::PERSISTENT_BUMP,
            lending_bump::PERSISTENT_BUMP,
        );
    }

    /// Get interest rate model for pool
    pub fn get_model(env: &Env, pool_id: &String) -> InterestRateModel {
        let key = LendingKey::InterestRateModel(pool_id.clone());
        if env.storage().persistent().has(&key) {
            env.storage().persistent().extend_ttl(
                &key,
                lending_bump::PERSISTENT_BUMP,
                lending_bump::PERSISTENT_BUMP,
            );
            env.storage().persistent().get(&key).unwrap_or_default()
        } else {
            InterestRateModel::default()
        }
    }

    /// Get accumulated interest index
    pub fn get_interest_index(env: &Env, pool_id: &String) -> i128 {
        let key = LendingKey::PoolInterestIndex(pool_id.clone());
        env.storage().persistent().get(&key).unwrap_or(PRECISION)
    }

    /// Set accumulated interest index
    pub fn set_interest_index(env: &Env, pool_id: &String, index: i128) {
        let key = LendingKey::PoolInterestIndex(pool_id.clone());
        env.storage().persistent().set(&key, &index);
        env.storage().persistent().extend_ttl(
            &key,
            lending_bump::PERSISTENT_BUMP,
            lending_bump::PERSISTENT_BUMP,
        );
    }

    /// Get last accrual timestamp
    pub fn get_last_accrual(env: &Env, pool_id: &String) -> u64 {
        let key = LendingKey::PoolLastAccrual(pool_id.clone());
        env.storage().persistent().get(&key).unwrap_or(0)
    }

    /// Set last accrual timestamp
    pub fn set_last_accrual(env: &Env, pool_id: &String, timestamp: u64) {
        let key = LendingKey::PoolLastAccrual(pool_id.clone());
        env.storage().persistent().set(&key, &timestamp);
        env.storage().persistent().extend_ttl(
            &key,
            lending_bump::PERSISTENT_BUMP,
            lending_bump::PERSISTENT_BUMP,
        );
    }

    /// Calculate new interest index with **compound** interest.
    ///
    /// Uses per-second compounding via iterative squaring (exponentiation by
    /// squaring) to compute: `new_index = current_index * (1 + rate_per_second) ^ time_elapsed`.
    ///
    /// This avoids the underestimation inherent in the linear approximation
    /// `index * (1 + rate * time)` for longer accrual intervals.
    /// Uses checked arithmetic to prevent overflow.
    pub fn calculate_new_index(current_index: i128, borrow_rate: i128, time_elapsed: u64) -> i128 {
        if time_elapsed == 0 {
            return current_index;
        }

        // rate_per_second in PRECISION units
        let rate_per_second = borrow_rate / (SECONDS_PER_YEAR as i128);

        // base = 1 + rate_per_second  (in PRECISION units)
        let base = PRECISION
            .checked_add(rate_per_second)
            .expect("Interest calculation base overflow");

        // Compute base^time_elapsed using exponentiation by squaring
        let compound_factor = Self::pow_precision(base, time_elapsed);

        current_index
            .checked_mul(compound_factor)
            .expect("Interest calculation index overflow")
            / PRECISION
    }

    /// Fixed-point exponentiation by squaring.
    ///
    /// Computes `base ^ exp` where `base` is in PRECISION units.
    /// Returns the result in PRECISION units.
    fn pow_precision(base: i128, exp: u64) -> i128 {
        if exp == 0 {
            return PRECISION;
        }

        let mut result = PRECISION;
        let mut b = base;
        let mut e = exp;

        while e > 0 {
            if e & 1 == 1 {
                result = (result * b) / PRECISION;
            }
            b = (b * b) / PRECISION;
            e >>= 1;
        }

        result
    }
}
